pkg/endpoint/policy.go

```diff
@@ -257,61 +155,53 @@ func (e *Endpoint) resolveL4Policy(owner Owner, repo *policy.Repository, c *poli
 	}
 
 	newL4EgressPolicy, err := repo.ResolveL4EgressPolicy(&egressCtx)
+
 	if err != nil {
 		return err
 	}
 
 	newL4Policy := &policy.L4Policy{Ingress: *newL4IngressPolicy,
 		Egress: *newL4EgressPolicy}
 
-	if reflect.DeepEqual(c.L4Policy, newL4Policy) {
+	if reflect.DeepEqual(e.DesiredL4Policy, newL4Policy) {
 		return nil
 	}
 
-	c.L4Policy = newL4Policy
+	e.DesiredL4Policy = newL4Policy
 	return nil
 }
 
-// Must be called with global endpoint.Mutex held
-// Returns a boolean to signalize if the policy was changed;
-// and a map matching which rules were successfully added/modified;
-// and a map matching which rules were successfully removed.
-// Must be called with Consumable mutex held.
-func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *identityPkg.IdentityCache,
-	repo *policy.Repository, c *policy.Consumable) (err error) {
-
-	// Mark all entries unused by denying them
-	for ingressIdentity := range c.IngressIdentities {
-		// Mark as false indicates denying
-		c.IngressIdentities[ingressIdentity] = false
-	}
-
-	for egressIdentity := range c.EgressIdentities {
-		c.EgressIdentities[egressIdentity] = false
-	}
-
-	// L4 policy needs to be applied on two conditions
-	// 1. The L4 policy has changed
-	// 2. The set of applicable security identities has changed.
-	if e.L4Policy != c.L4Policy || e.LabelsMap != labelsMap {
-
-		// PolicyMap can't be created in dry mode.
-		if !owner.DryModeEnabled() {
-			// Collect unused redirects.
-			err = e.applyL4PolicyLocked(e.LabelsMap, labelsMap, e.L4Policy, c.L4Policy)
-			if err != nil {
-				// This should not happen, and we can't fail at this stage anyway.
-				e.getLogger().WithError(err).Error("L4 Policy application failed")
-				return
-			}
+func (e *Endpoint) computeDesiredPolicyMapState(owner Owner, labelsMap *identityPkg.IdentityCache,
+	repo *policy.Repository) {
+	desiredPolicyKeys := make(map[policymap.PolicyKey]struct{})
+	if e.LabelsMap != labelsMap {
+		e.LabelsMap = labelsMap
+	}
+	e.computeDesiredL4PolicyMapEntries(desiredPolicyKeys)
+	e.determineAllowLocalhost(owner, desiredPolicyKeys)
+	e.computeDesiredL3PolicyMapEntries(owner, labelsMap, repo, desiredPolicyKeys)
+	e.desiredMapState = desiredPolicyKeys
+}
+
+func (e *Endpoint) determineAllowLocalhost(owner Owner, desiredPolicyKeys map[policymap.PolicyKey]struct{}) {
+
+	if desiredPolicyKeys == nil {
```

Comment:
This was moreso done out of caution than anything else. 
I'll add a nil check for `e.DeisredL4Policy`.

I wanted a smaller function because it makes the code more readable.

Reply:
We could risk it. There is a definite order of operations to these functions, so it might never happen. An alternative is to create all these maps as empties in the NewEndpoint functions. I'm just not sure if we would then try to effect an empty policy somewhere, before we read all the policies from k8s etc.